### PR TITLE
8315024: Vector API FP reduction tests should not test for exact equality

### PR DESCRIPTION
**Pull request by amosshi**
Backport of [JDK-8315024](https://bugs.openjdk.org/browse/JDK-8315024)
- This PR has two commits
  - `commit 1`: the git apply result, which is `clean`
  - `commit 2`: manually merge the following two `.rej` files
    - `test/jdk/jdk/incubator/vector/DoubleMaxVectorTests.java.rej`
    - `test/jdk/jdk/incubator/vector/FloatMaxVectorTests.java.rej`
  - `commit 3,4`: Fix code postion
  - `commit 5`: Add file `Unit-header.template` 

Content of `test/jdk/jdk/incubator/vector/DoubleMaxVectorTests.java.rej`

```diff
@@ -668,15 +670,21 @@
 
     static void assertReductionArraysEquals(double[] r, double rc, double[] a,
                                             FReductionOp f, FReductionAllOp fa) {
+        assertReductionArraysEquals(r, rc, a, f, fa, (double)0.0);
+    }
+
+    static void assertReductionArraysEquals(double[] r, double rc, double[] a,
+                                            FReductionOp f, FReductionAllOp fa,
+                                            double relativeError) {
         int i = 0;
         try {
-            Assert.assertEquals(rc, fa.apply(a));
+            Assert.assertEquals(rc, fa.apply(a), Math.abs(rc * relativeError));
             for (; i < a.length; i += SPECIES.length()) {
-                Assert.assertEquals(r[i], f.apply(a, i));
+                Assert.assertEquals(r[i], f.apply(a, i), Math.abs(r[i] * relativeError));
             }
         } catch (AssertionError e) {
-            Assert.assertEquals(rc, fa.apply(a), "Final result is incorrect!");
-            Assert.assertEquals(r[i], f.apply(a, i), "at index #" + i);
+            Assert.assertEquals(rc, fa.apply(a), Math.abs(rc * relativeError), "Final result is incorrect!");
+            Assert.assertEquals(r[i], f.apply(a, i), Math.abs(r[i] * relativeError), "at index #" + i);
         }
     }
 
@@ -690,15 +698,22 @@
 
     static void assertReductionArraysEqualsMasked(double[] r, double rc, double[] a, boolean[] mask,
                                             FReductionMaskedOp f, FReductionAllMaskedOp fa) {
+        assertReductionArraysEqualsMasked(r, rc, a, mask, f, fa, (double)0.0);
+    }
+
+    static void assertReductionArraysEqualsMasked(double[] r, double rc, double[] a, boolean[] mask,
+                                            FReductionMaskedOp f, FReductionAllMaskedOp fa,
+                                            double relativeError) {
         int i = 0;
         try {
-            Assert.assertEquals(rc, fa.apply(a, mask));
+            Assert.assertEquals(rc, fa.apply(a, mask), Math.abs(rc * relativeError));
             for (; i < a.length; i += SPECIES.length()) {
-                Assert.assertEquals(r[i], f.apply(a, i, mask));
+                Assert.assertEquals(r[i], f.apply(a, i, mask), Math.abs(r[i] *
+relativeError));
             }
         } catch (AssertionError e) {
-            Assert.assertEquals(rc, fa.apply(a, mask), "Final result is incorrect!");
-            Assert.assertEquals(r[i], f.apply(a, i, mask), "at index #" + i);
+            Assert.assertEquals(rc, fa.apply(a, mask), Math.abs(rc * relativeError), "Final result is incorrect!");
+            Assert.assertEquals(r[i], f.apply(a, i, mask), Math.abs(r[i] * relativeError), "at index #" + i);
         }
     }
```

Content of `test/jdk/jdk/incubator/vector/FloatMaxVectorTests.java.rej`

```diff
@@ -668,15 +670,21 @@
 
     static void assertReductionArraysEquals(float[] r, float rc, float[] a,
                                             FReductionOp f, FReductionAllOp fa) {
+        assertReductionArraysEquals(r, rc, a, f, fa, (float)0.0);
+    }
+
+    static void assertReductionArraysEquals(float[] r, float rc, float[] a,
+                                            FReductionOp f, FReductionAllOp fa,
+                                            float relativeError) {
         int i = 0;
         try {
-            Assert.assertEquals(rc, fa.apply(a));
+            Assert.assertEquals(rc, fa.apply(a), Math.abs(rc * relativeError));
             for (; i < a.length; i += SPECIES.length()) {
-                Assert.assertEquals(r[i], f.apply(a, i));
+                Assert.assertEquals(r[i], f.apply(a, i), Math.abs(r[i] * relativeError));
             }
         } catch (AssertionError e) {
-            Assert.assertEquals(rc, fa.apply(a), "Final result is incorrect!");
-            Assert.assertEquals(r[i], f.apply(a, i), "at index #" + i);
+            Assert.assertEquals(rc, fa.apply(a), Math.abs(rc * relativeError), "Final result is incorrect!");
+            Assert.assertEquals(r[i], f.apply(a, i), Math.abs(r[i] * relativeError), "at index #" + i);
         }
     }
 
@@ -690,15 +698,22 @@
 
     static void assertReductionArraysEqualsMasked(float[] r, float rc, float[] a, boolean[] mask,
                                             FReductionMaskedOp f, FReductionAllMaskedOp fa) {
+        assertReductionArraysEqualsMasked(r, rc, a, mask, f, fa, (float)0.0);
+    }
+
+    static void assertReductionArraysEqualsMasked(float[] r, float rc, float[] a, boolean[] mask,
+                                            FReductionMaskedOp f, FReductionAllMaskedOp fa,
+                                            float relativeError) {
         int i = 0;
         try {
-            Assert.assertEquals(rc, fa.apply(a, mask));
+            Assert.assertEquals(rc, fa.apply(a, mask), Math.abs(rc * relativeError));
             for (; i < a.length; i += SPECIES.length()) {
-                Assert.assertEquals(r[i], f.apply(a, i, mask));
+                Assert.assertEquals(r[i], f.apply(a, i, mask), Math.abs(r[i] *
+relativeError));
             }
         } catch (AssertionError e) {
-            Assert.assertEquals(rc, fa.apply(a, mask), "Final result is incorrect!");
-            Assert.assertEquals(r[i], f.apply(a, i, mask), "at index #" + i);
+            Assert.assertEquals(rc, fa.apply(a, mask), Math.abs(rc * relativeError), "Final result is incorrect!");
+            Assert.assertEquals(r[i], f.apply(a, i, mask), Math.abs(r[i] * relativeError), "at index #" + i);
         }
     }
```

Testing
- Local: Test passed on `MacOS 14.5` with `Apple M1 Max`
  - `Double128VectorTests.java`: Test results: passed: 1
  - `Double256VectorTests.java`: Test results: passed: 1
  - `Double512VectorTests.java`: Test results: passed: 1
  - `Double64VectorTests.java`: Test results: passed: 1
  - `DoubleMaxVectorTests.java`: Test results: passed: 1
  - `Float128VectorTests.java`: Test results: passed: 1
  - `Float256VectorTests.java`: Test results: passed: 1
  - `Float512VectorTests.java`: Test results: passed: 1
  - `Float64VectorTests.java`: Test results: passed: 1
  - `FloatMaxVectorTests.java`: Test results: passed: 1
- Pipeline:
  - Linux, Windows - Passed
  - MacOS - Failed with message: `This request was automatically failed because there were no enabled runners online to process the request for more than 1 days.`
- Testing Machine: SAP nightlies SUCCESSFUL on `2024-08-17`
  - Automated jtreg test: `jtreg_jdk_tier3`, Started at `2024-08-16 22:45:27+01:00`
  - jdk/incubator/vector/Double128VectorTests.java: SUCCESSFUL GitHub 📊 - [22:55:08.718 -> 61,687 msec]
  - jdk/incubator/vector/Double256VectorTests.java: SUCCESSFUL GitHub 📊 - [22:55:09.720 -> 52,033 msec]
  - jdk/incubator/vector/Double512VectorTests.java: SUCCESSFUL GitHub 📊 - [22:55:25.860 -> 47,865 msec]
  - jdk/incubator/vector/Double64VectorTests.java: SUCCESSFUL GitHub 📊 - [22:55:31.432 -> 79,579 msec]
  - jdk/incubator/vector/DoubleMaxVectorTests.java: SUCCESSFUL GitHub 📊 - [22:55:40.700 -> 44,241 msec]
  - jdk/incubator/vector/Float128VectorTests.java: SUCCESSFUL GitHub 📊 - [22:55:55.119 -> 51,644 msec]
  - jdk/incubator/vector/Float256VectorTests.java: SUCCESSFUL GitHub 📊 - [22:56:00.304 -> 46,476 msec]
  - jdk/incubator/vector/Float512VectorTests.java: SUCCESSFUL GitHub 📊 - [22:56:01.753 -> 49,030 msec]
  - jdk/incubator/vector/Float64VectorTests.java: SUCCESSFUL GitHub 📊 - [22:56:03.088 -> 75,969 msec]
  - jdk/incubator/vector/FloatMaxVectorTests.java: SUCCESSFUL GitHub 📊 - [22:56:10.405 -> 45,711 msec]

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] [JDK-8315024](https://bugs.openjdk.org/browse/JDK-8315024) needs maintainer approval

### Issue
 * [JDK-8315024](https://bugs.openjdk.org/browse/JDK-8315024): Vector API FP reduction tests should not test for exact equality (**Bug** - P5 - Approved)

### Reviewers
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk17u-dev.git pull/2768/head:pull/2768` \
`$ git checkout pull/2768`

Update a local copy of the PR: \
`$ git checkout pull/2768` \
`$ git pull https://git.openjdk.org/jdk17u-dev.git pull/2768/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 2768`

View PR using the GUI difftool: \
`$ git pr show -t 2768`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk17u-dev/pull/2768.diff">https://git.openjdk.org/jdk17u-dev/pull/2768.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk17u-dev/pull/2768#issuecomment-2263567064)

### CODE REVIEW

**Comment by MBaesken** [comment]
In the original commit I see  in [test/jdk/jdk/incubator/vector/DoubleMaxVectorTests.java](https://github.com/openjdk/jdk/commit/e6f23a90d4a53339a3c9c2b76fc5d317940e4472#diff-a1baf5cd2c3bc9046a703b75c78cdb1948a2ca10d84ac1428dee4b471f87546f)

https://github.com/openjdk/jdk/commit/e6f23a90d4a53339a3c9c2b76fc5d317940e4472
private static final double RELATIVE_ROUNDING_ERROR = (double)0.000001;

Is it intentional that it is not in the backport ?

**Comment by amosshi** [comment]
Hi @MBaesken,

Somehow the git `patch` command has put these two lines to the `test/jdk/jdk/incubator/vector/DoubleMaxVectorTests.java` file at line `625-626`

```diff
+ 625    // for floating point reduction ops that may introduce rounding errors
+ 626    private static final double RELATIVE_ROUNDING_ERROR = (double)0.000001;
``` 

I will move them to earlier lines in respect to the original commit.


Regards
Amos
A Happy Developer

**Comment by amosshi** [comment]
### Update
- Sending to test server to test again

File: test/jdk/jdk/incubator/vector/DoubleMaxVectorTests.java [hunk 1]
```diff
@@ -609,7 +625,6 @@ static void assertDoubleBroadcastArraysEquals(double[] r, double[] a, double[] b
         }
     }
 
-
```
Comment (TheRealMDoerr) [inline test/jdk/jdk/incubator/vector/DoubleMaxVectorTests.java:612]: This empty line should not get removed.

File: test/jdk/jdk/incubator/vector/FloatMaxVectorTests.java [hunk 2]
```diff
@@ -609,7 +625,6 @@ static void assertDoubleBroadcastArraysEquals(float[] r, float[] a, float[] b, f
         }
     }
 
-
```
Comment (TheRealMDoerr) [inline test/jdk/jdk/incubator/vector/FloatMaxVectorTests.java:612]: This empty line should not get removed.

File: test/jdk/jdk/incubator/vector/templates/Unit-header.template [hunk 3]
```diff
@@ -157,7 +164,25 @@ public class $vectorteststype$ extends AbstractVectorTest {
...
+        try {
+            Assert.assertEquals(rc, fa.apply(a), Math.abs(rc * relativeError));
+            for (; i < a.length; i += SPECIES.length()) {
+                Assert.assertEquals(r[i], f.apply(a, i), Math.abs(r[i] * relativeError));
+            }
+        } catch (AssertionError e) {
+            Assert.assertEquals(rc, fa.apply(a), Math.abs(rc * relativeError), "Final result is incorrect!");
+            Assert.assertEquals(r[i], f.apply(a, i), Math.abs(r[i] * relativeError), "at index #" + i);
+        }
+     }
```
Comment (TheRealMDoerr) [inline test/jdk/jdk/incubator/vector/templates/Unit-header.template:184]: 1 space too much.

File: test/jdk/jdk/incubator/vector/templates/Unit-header.template [hunk 4]
```diff
@@ -179,7 +207,26 @@ public class $vectorteststype$ extends AbstractVectorTest {
...
+            Assert.assertEquals(rc, fa.apply(a, mask), Math.abs(rc * relativeError));
+            for (; i < a.length; i += SPECIES.length()) {
+                Assert.assertEquals(r[i], f.apply(a, i, mask), Math.abs(r[i] *
+relativeError));
+            }
+        } catch (AssertionError e) {
+            Assert.assertEquals(rc, fa.apply(a, mask), Math.abs(rc * relativeError), "Final result is incorrect!");
+            Assert.assertEquals(r[i], f.apply(a, i, mask), Math.abs(r[i] * relativeError), "at index #" + i);
+        }
+     }
```
Comment (TheRealMDoerr) [inline test/jdk/jdk/incubator/vector/templates/Unit-header.template:228]: 1 space too much.

**Review by TheRealMDoerr — COMMENTED** [review]
Looks basically good. I only found a couple of formatting differences which should get removed.

File: test/jdk/jdk/incubator/vector/DoubleMaxVectorTests.java (hunk 1 above)
Comment (amosshi) [reply test/jdk/jdk/incubator/vector/DoubleMaxVectorTests.java]: Fixed.
It is indeed there are 2 empty lines at https://github.com/openjdk/jdk/blob/e6f23a90d4a53339a3c9c2b76fc5d317940e4472/test/jdk/jdk/incubator/vector/DoubleMaxVectorTests.java#L720-L721

File: test/jdk/jdk/incubator/vector/FloatMaxVectorTests.java (hunk 2 above)
Comment (amosshi) [reply test/jdk/jdk/incubator/vector/FloatMaxVectorTests.java]: Fixed.
Now it is in line with https://github.com/openjdk/jdk/blob/e6f23a90d4a53339a3c9c2b76fc5d317940e4472/test/jdk/jdk/incubator/vector/FloatMaxVectorTests.java#L720-L721

File: test/jdk/jdk/incubator/vector/templates/Unit-header.template (hunk 3 above)
Comment (amosshi) [reply test/jdk/jdk/incubator/vector/templates/Unit-header.template]: Fixed.
Extra space removed.

File: test/jdk/jdk/incubator/vector/templates/Unit-header.template (hunk 4 above)
Comment (amosshi) [reply test/jdk/jdk/incubator/vector/templates/Unit-header.template]: Fixed. 
1 extra space removed.

**Review by TheRealMDoerr — APPROVED** [review]
Thanks for the update!

**Comment by amosshi** [comment]
/approval request "Backporting for parity with 17.0.13-oracle. No risk, test fix. Un-Clean backport. Test passes and SAP nightlies pass on 2024-08-17”

**Pull request closed without merging**